reana-ui/src/pages/badgeCreator/LauncherBadgeCreator.js

```diff
@@ -0,0 +1,297 @@
+
+  const handleDeleteParam = (index, paramKey) => {
+    // Clear the validation error relative to the parameter that has been deleted
+    setValidationErrors((prevErrors) => ({
+      ...prevErrors,
+      parameters: { ...prevErrors.parameters, [paramKey]: null },
+    }));
+    setLauncherData((prevData) => {
+      const newParameters = [...prevData.parameters];
+      newParameters.splice(index, 1);
+      return { ...prevData, parameters: newParameters };
+    });
+  };
+
+  const handleParamChange = (index, key, value, paramKey) => {
+    // Clear the validation errors relative to the parameter that has been changed
+    setValidationErrors((prevErrors) => ({
+      ...prevErrors,
+      parameters: { ...prevErrors.parameters, [paramKey]: null },
+    }));
+    setLauncherData((prevData) => {
+      const newParameters = [...prevData.parameters];
+      newParameters[index] = { ...newParameters[index], [key]: value };
+      return { ...prevData, parameters: newParameters };
+    });
+  };
+
+  const handleSubmit = () => {
+    let errorMessage = null;
+    let paramsString = null;
+    // Validate URL
+    try {
+      new URL(launcherData.url);
+    } catch (e) {
+      errorMessage = "The form contains invalid data!";
+      setValidationErrors((prevErrors) => ({
+        ...prevErrors,
+        url: "Not a valid URL!",
+      }));
+    }
+
+    // Validate analysis name
+    if (launcherData.analysisName) {
+      if (launcherData.analysisName.includes(".")) {
+        errorMessage = "The form contains invalid data!";
+        setValidationErrors((prevErrors) => ({
+          ...prevErrors,
+          analysisName: "The analysis name cannot contain dots!",
+        }));
+      }
+    }
+
+    // Validate specification file name
+    if (
+      launcherData.specFileName &&
+      launcherData.specFileName !== "reana.yaml"
+    ) {
+      if (
+        !launcherData.specFileName.endsWith(".yaml") &&
+        !launcherData.specFileName.endsWith(".yml")
+      ) {
+        errorMessage = "The form contains invalid data!";
+        setValidationErrors((prevErrors) => ({
+          ...prevErrors,
+          specFileName:
+            "The specification file name must end with .yaml or .yml",
+        }));
+      }
+    }
+
+    if (launcherData.parameters.length > 0) {
+      // The value of each parameter is considered to be a JSON-encoded string.
+      // In this way, the user can pass integers, lists, complex objects, etc.
+
+      // Validate the value of each parameter (valid JSON string)
+      let paramsObject = {};
+      launcherData.parameters.forEach((param, index) => {
+        try {
+          paramsObject[param.key] = JSON.parse(param.value);
+        } catch (e) {
+          errorMessage = "The form contains invalid data!";
+          setValidationErrors((prevErrors) => ({
+            ...prevErrors,
+            parameters: {
+              ...prevErrors.parameters,
+              [param.key]: "Not a valid JSON string!",
```

Comment:
We are missing a couple of checks for parameters:
1. keys must not be empty
2. keys must be unique

Also note that, if two keys are the same (or if both are empty), then the errors will be shown for both input fields, even if one of them has no validation issues. Maybe we could access/store errors by index, instead of by key, to avoid this?

Reply:
Done! Now the error could also be displayed on the key form, in case it's duplicate or not unique. I think all the cases should be handled now, there is just one exception:
1. Insert 2 equal keys
2. Try to create the badge
3. Get the "duplicate" error on the _second_ key
4. Delete / change the _first_ key
Now the second key still displays the "duplicate key" error, because its validation state has not been cleaned. Of course, after changing that key or trying to create the badge, the error will disappear.

I don't think that's a big deal, and adding the code to gracefully handle this edge case is not really worth it. What do you think?